src/compiler/types.ts

```diff
@@ -603,15 +603,75 @@ namespace ts {
 
     export type DeclarationName = Identifier | StringLiteral | NumericLiteral | ComputedPropertyName | BindingPattern;
 
-    export interface RealDeclaration extends Node {
+    export interface DeclarationBase extends Node {
         _declarationBrand: any;
         name?: DeclarationName;
     }
 
-    // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
-    export type Declaration = RealDeclaration | BinaryExpression;
-
-    export interface DeclarationStatement extends RealDeclaration, Statement {
+    export type Declaration =
+        | ArrowFunction
+        // Binary expressions can be declarations if they are 'exports.foo = bar' expressions in JS files
+        | BinaryExpression

```

Comment:
What do you think about:
```ts
    export interface JsExportsIdentifier extends Identifier {
        text: "exports";
    }

    export interface JsExportsPropertyAccess extends PropertyAccessExpression {
        expression: JsExportsIdentifier;
    }

    export interface JsExportsElementAccess extends ElementAccessExpression {
        expression: JsExportsIdentifier;
    }

    export interface JsExportExpression extends AssignmentExpression<Token<SyntaxKind.EqualsToken>>, Declaration {
        left: JsExportsPropertyAccess | JsExportsElementAccess;
    }
```

Then we have a type that represents a specific shape for a `BinaryExpression` that represents a declaration, rather than a large discriminated union?